platform/centec/sonic-platform-modules-e582/48x2q4z/modules/centec_at24c64.c

```diff
@@ -1,657 +1,602 @@
+            dev_warn(&client->dev, "cannot write due to controller restrictions.");
+        }
+    }
+
+    at24->client[0] = client;
+
+    /* use dummy devices for multiple-address chips */
+    for (i = 1; i < num_addresses; i++) {
+        at24->client[i] = i2c_new_dummy(client->adapter, client->addr + i);
+        if (!at24->client[i]) {
+            dev_err(&client->dev, "address 0x%02x unavailable\n", client->addr + i);
+            err = -EADDRINUSE;
+            goto err_clients;
+        }
+    }
+
+    err = sysfs_create_bin_file(&client->dev.kobj, &at24->bin);
+    if (err)
+        goto err_clients;
+
+    i2c_set_clientdata(client, at24);
+
+    printk(KERN_ALERT "at24_probe %s done\n", client->name); 
+
+    return 0;
+
+err_clients:
+    for (i = 1; i < num_addresses; i++)
+        if (at24->client[i])
+            i2c_unregister_device(at24->client[i]);
+
+    return err;
+}
+
+static int at24_remove(struct i2c_client *client)
+{
+    struct at24_data *at24;
+    int i;
+
+    at24 = i2c_get_clientdata(client);
+    sysfs_remove_bin_file(&client->dev.kobj, &at24->bin);
+
+    for (i = 1; i < at24->num_addresses; i++)
+        i2c_unregister_device(at24->client[i]);
+
+    return 0;
+}
+
+/*-------------------------------------------------------------------------*/
+
+static struct i2c_board_info i2c_devs = {
+    I2C_BOARD_INFO("24c64-ctc", 0x57),
+};
+
+static struct i2c_adapter *adapter = NULL;
+static struct i2c_client  *client  = NULL;
+
+static int ctc_at24c64_init(void)
+{
+    printk(KERN_ALERT "ctc_at24c64_init\n");
+
+    adapter = i2c_get_adapter(0);
+    if(adapter == NULL){
+        printk(KERN_ALERT "i2c_get_adapter == NULL\n");
+        return -1;
+    }
+
+    client = i2c_new_device(adapter, &i2c_devs);
+    if(client == NULL){
+        printk(KERN_ALERT "i2c_new_device == NULL\n");
+        i2c_put_adapter(adapter);
+        adapter = NULL;
+        return -1;
+    }
+
+    return 0;
+}
+
+static void ctc_at24c64_exit(void)
+{
+    printk(KERN_ALERT "ctc_at24c64_exit\n");
+    if(client){
+        i2c_unregister_device(client);
+    }
+    if(adapter){
+        i2c_put_adapter(adapter);
+    }
+} 
+
+static struct i2c_driver at24_ctc_driver = {
+    .driver = {
+        .name = "at24-ctc",
+        .owner = THIS_MODULE,
+    },
+    .probe = at24_probe,
+    .remove = at24_remove,
+    .id_table = at24_ctc_ids,
+};
+
+static int __init at24_ctc_init(void)
+{
+    if (!io_limit) {
+        pr_err("at24_ctc: io_limit must not be 0!\n");
+        return -EINVAL;
+    }
+
+    io_limit = rounddown_pow_of_two(io_limit);
+    
+    ctc_at24c64_init();
+    
+    return i2c_add_driver(&at24_ctc_driver);
+}
+module_init(at24_ctc_init);
+
+static void __exit at24_ctc_exit(void)
+{
+    ctc_at24c64_exit();
+    i2c_del_driver(&at24_ctc_driver);
+}
+module_exit(at24_ctc_exit);
+
+MODULE_DESCRIPTION("Driver for most I2C EEPROMs");
+MODULE_AUTHOR("David Brownell and Wolfram Sang");
+MODULE_LICENSE("GPL");
+/* XXX */
+
```

Comment:
why not use standard at24 kernel driver?

Reply:
hi lguohan:
thanks for review
i created a eeprom sys file in private driver for user to read and get eeprom data
/sys/class/i2c-adapter/i2c-0/0-0057/eeprom this file is created in this private driver

i tried to use ac24 kernel standard driver, could not found eeprom sys file to read eeprom data, did i miss something, please help me how to do it with standard at24 driver?

i see 
 sonic-buildimage/platform/broadcom/sonic-platform-modules-cel/dx010/modules/mc24lc64t.c
use the same way as i do